### PR TITLE
Let users define flash chip specs

### PR DESCRIPTION
**Pull request by jflyper**
This small PR should eliminate support requests for different flash chips.

### CODE REVIEW

**Comment by jflyper** [comment]
I have a couple of question. This PR has four failed targets in Travis CI.

1. 
Three are package downloading failures during environment building. Can I expect these go away on the next PR push?
2. 
Another failed target is the ALIENFLIGHTF1, with 20 bytes overflow in FLASH. This is target is said to be almost identical to NAZE, but the NAZE build is more than 500 bytes smaller than ALIENFLIGHTF1. Do some one know where the extra 500 bytes came from?

**Comment by martinbudden** [comment]
@jflyper 
1. normally the download failures go away on the next push.
2. ALIENFLIGHTF1 includes a custom octocopter mixer, that probably accounts for most of the 500 bytes

**Comment by jflyper** [comment]
@martinbudden 
Thanks.

**Comment by jflyper** [comment]
The new feature is now an option CUSTOM_FLASHCHIP, and not included in any target for now (no wonder it passes the check!) If someone cares about the extension, then handling of the option should be reconsidered.

**Comment by MJ666** [comment]
It wold be posible to undefine a few features for the AlienFlight F1 target (i.e. softseria portsl, gps, 4wire etc.). @martinbudden looks to be the changed custom mixer code is taking more space in the flash as before the recent changes. Very strange.

File: src/main/config/config_system.h [hunk 1]
```diff
@@ -20,6 +20,11 @@
 typedef struct systemConfig_s {
     uint8_t emf_avoidance;                   // change pll settings to avoid noise in the uhf band
     uint8_t i2c_highspeed;                   // Overclock i2c Bus for faster IMU readings
+#ifdef CUSTOM_FLASHCHIP
+    uint32_t flashchip_id;
+    uint16_t flashchip_nsect;
+    uint16_t flashchip_pps;
+#endif
 } systemConfig_t;
```
Comment (hydra) [inline src/main/config/config_system.h]: this doesn't belong here.  create a new parameter group specifically for it.

https://en.wikipedia.org/wiki/Separation_of_concerns

Developers shouldn't have to change system configuration code to make changes to flash chip configuration information.

**Comment by hydra** [comment]
yes i like the idea, users can then do what they want and manufactuers of boards can tell users what to configure.

i think however there should also be a corresponding MSP command to configure the flash chip rather than request the users do it via the CLI.  See #922 

A long term goal is to allow target specification via the configurator and not via the firmware.  This seems like a step towards that goal to me.  Think picture of board and a pop-up for blackbox/flash hardware configuration.

File: src/main/config/config_system.h (hunk 1 above)
Comment (jflyper) [inline src/main/config/config_system.h]: Made a new pg called PG_DRIVER_FLASHCHIP_CONFIG.

**Comment by jflyper** [comment]
@hydra Yes, I think I shared your vision of "target specification via configurator". I think it's awesome. Let me see what I can do for the MSP support.

**Comment by jflyper** [comment]
Do we have generic MSP functionalities that deals with PGs?

**Comment by ledvinap** [comment]
@jflyper : There is some (proof-of-concept) code here: https://github.com/cleanflight/cleanflight/blob/master/src/main/io/msp.c#L462

**Comment by jflyper** [comment]
@ledvinap 
I've took a look at the PoC code. So it's a general purpose MSP-PG two way mapping service, correct?

I was imagining something like a PG manipulation facility/protocol running on top of MSP. There will be only two MSP commands, MSP_GET_PG and MSP_SET_PG. Payloads are identified with PG ids. Just like a simple RPC. Some sort of data serialization/deserialization service will be nice. Thoughts?

**Comment by ledvinap** [comment]
@jflyper : Yes, it makes sense, but is not implemented yet.
Some thought should be given to record versioning/profile/array handling etc, so the protocol is not obsolete too soon.
Implementing this will allow full backup/restore from CF configurator without using CLI ...

**Comment by jflyper** [comment]
@ledvinap 
> Yes, it makes sense, but is not implemented yet.
Some thought should be given to record versioning/profile/array handling etc, so the protocol is not obsolete too soon.
Implementing this will allow full backup/restore from CF configurator without using CLI ...

Makes tons of sense. I'm in.

**Comment by jflyper** [comment]
@ledvinap 
Haven't tested the SET side yet, but the GET side worked like a charm.
Two defs + one map entry = Done. Super.

**Comment by jflyper** [comment]
@ledvinap 
SET side verified for correct operation also.

File: src/main/io/msp_protocol.h [hunk 2]
```diff
@@ -164,6 +164,8 @@
 
 #define MSP_TRANSPONDER_CONFIG          82 //out message         Get transponder settings
 #define MSP_SET_TRANSPONDER_CONFIG      83 //in  message         Set transponder settings
+#define MSP_FLASHCHIP_CONFIG            84 //out  message        Get flash chip settings
```
Comment (gaelj) [inline src/main/io/msp_protocol.h]: IDs 84 & 85 are used already by https://github.com/cleanflight/cleanflight/pull/2112/files.
I also used 86 & 87 in https://github.com/cleanflight/cleanflight/pull/2173/commits/14471c16ff3a8bc14af3fa36eddee09b0cbe3a19 (merged)
88 & 89 should be available.

Comment (jflyper) [inline src/main/io/msp_protocol.h]: Got it.

**Comment by jflyper** [comment]
ALIENFLIGHTF1 target will have NO_CUSTOM_FLASHCHIP defined now.

**Comment by hydra** [comment]
ok the MSP commands have been dropped, we can wait for PG <id> GET/SET code as required.  Merged in the feature and the CLI changes.   I'm only enabling this feature for boards that have space for user-solderable flash chips since they are the only users that should be needing the feature - users of clone boards with different flash chips can buy original boards if they want support or compile it themselves.

**Comment by hydra** [comment]
merged

**Pull request closed without merging**